## BALDR

BALDR is a pipeline for reconstructing human or rhesus macaque immunoglobulin(Ig)/B cell receptor(BCR) sequences from single cell RNA-Seq data generated by Illumina sequencing. 

BALDR is based on the *de novo* assembly of RNA-Seq reads. It allows reconstructions using following methods:
1. IG-mapped_Unmapped (human and rhesus) - Assemble reads mapping to Ig loci + Unmapped reads [default for human]
2. FilterNonIG (rhesus)- Assemble reads after filtering those that match to non-Ig genes in the genome [default for rhesus]
3. Unfiltered (human and rhesus) - Use all reads for assembling transcripts and select Ig transcript models
4. IG-mapped_only (human and rhesus) - Assemble reads mapping to Ig loci
5. Recombinome-mapped (human) - Assemble reads mapping to the Ig recombinome
6. IMGT-mapped (human) - Assemble reads mapping to IMGT V(D)J and C sequences

## Installation (manual)

### Pre-requisites
1. [Trimmomatic 0.32](http://www.usadellab.org/cms/?page=trimmomatic)
2. [Trinity v2.3.2](https://github.com/trinityrnaseq/trinityrnaseq/wiki) (Newer versions are not compatible)
3. [bowtie2 2.3.0](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml)
4. [STAR v2.5.2b](https://github.com/alexdobin/STAR)
5. [samtools v1.3.1](http://www.htslib.org/download/)
6. [IgBLAST v1.6.1](https://www.ncbi.nlm.nih.gov/igblast/faq.html#standalone) (Newer versions are not compatible)
7. [seqtk 1.2](https://github.com/lh3/seqtk)
8. Perl 5

Install all the pre-requisites
Clone or download the BALDR package. 
```
cd BALDR
chmod +x BALDR
```
## Docker image

You can pull the image from DockerHub
```
docker pull bosingerlab/docker
```

OR

You can use the Dockerfile to build an image. 
```
mkdir baldr_docker
cd baldr_docker
wget https://raw.githubusercontent.com/BosingerLab/BALDR/master/Dockerfile
docker build -t bosingerlab/baldr .
```

## Running BALDR

### STAR genome index

If the --STAR_index flag is not provided, BALDR will download and generate the genome index. This takes time and if a docker container is used, it may end up using a lot more space for the docker image. It is thus recommended that the STAR genome index be generated before running BALDR. A pre-built index can also be loaded into the shared memory which will significantly decrease the runtime. 

Human
```
mkdir -p STAR_GRCh38/STAR_GRCh38_index
cd STAR_GRCh38
wget ftp://ftp.ensembl.org/pub/release-86/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz
wget ftp://ftp.ensembl.org/pub/release-86/gtf/homo_sapiens/Homo_sapiens.GRCh38.86.gtf.gz
gunzip *.gz
STAR --runThreadN 16 --runMode genomeGenerate --genomeDir STAR_GRCh38_index --genomeFastaFiles Homo_sapiens.GRCh38.dna.primary_assembly.fa --sjdbGTFfile Homo_sapiens.GRCh38.86.gtf --sjdbOverhang 100
```

Rhesus macaque
```
mkdir -p STAR_MacaM/STAR_MacaM_index
cd STAR_MacaM
wget https://www.unmc.edu/rhesusgenechip/MacaM_Rhesus_Genome_v7.fasta.bz2
bunzip2 $BALDR_path/resources/STAR_genomes/rhesus_monkey/MacaM_Rhesus_Genome_v7.fasta.bz2
wget https://www.unmc.edu/rhesusgenechip/RhesusGenomeUpload/MacaM_Rhesus_Genome_Annotation_v7.8.2.gtf
STAR --runThreadN 16 --runMode genomeGenerate --genomeDir STAR_MacaM_index --genomeFastaFiles MacaM_Rhesus_Genome_v7.fasta --sjdbGTFfile MacaM_Rhesus_Genome_Annotation_v7.8.2.gtf --sjdbOverhang 100
```

### Loading genome index in the shared memory

Before running BALDR, it is recommended to load the genome index in the memory
```
STAR --genomeDir </path/to/STAR/index/folder> --genomeLoad LoadAndExit
```

When BALDR has been run for all cells, run the following command to clear the genome index from the shared memory
```
STAR --genomeDir </path/to/STAR/index/folder> --genomeLoad Remove
```

## Command line usage
```
Single-end:
./BALDR --single <file.fastq.gz> <options>

Paired-end:
./BALDR --paired <R1.fastq.gz,R2.fastq.gz> <options>

Options:
  --trimmomatic  Path for trimmomatic.jar file (e.g. ~/Trimmomatic-0.36/trimmomatic-0.36.jar) (required)
  --adapter      Path for the Trimmomatic adapter file (e.g. ~/Trimmomatic-0.36/adapters/NexteraPE-PE.fa) (required)
  --trinity      Path for Trinity (e.g. ~/trinityrnaseq-Trinity-v2.3.2/Trinity) (required)
  --STAR         Path for the STAR binary (required)
  --BALDR        Path for the BALDR directory (e.g. ~/BALDR) (required)
  --method       One or more reconstruction methods. For multiple methods, separate only by comma
  	    	 human: IG-mapped_Unmapped (default)/Unfiltered/IG-mapped_only/IMGT-mapped/Recombinome-mapped 
  		 rhesus_monkey: FilterNonIG (default)/IG-mapped_Unmapped/Unfiltered/IG-mapped_only
  --organism     human (default) or rhesus_monkey
  --STAR_index   Path for the STAR aligner genome index
  --memory       Max memory for Trinity (default 32G)
  --threads      number of threads for STAR/bowtie2/Trinity (default 1)
  --version      Version
  --help         Print this help
```

## Running using the Docker image

```

```

## Output directories

The output is written in the working directory. The following folders are created:



## Aggregate results for all cells
... in progress

## Clonal assignment
... in progress

